### PR TITLE
upgrade gradle; change dependencies around

### PR DESCRIPTION
**Pull request by catoverdrive**
a few things happening here, most of which was me trying to not have to explicitly list dependencies in the shadowJar/shadowTestJar tasks:
- upgraded gradle to 5.0 and some plugins to be compatible
- split compile dependencies into "bundled" and "unbundled" to more explicitly separate the things we want in the jars and dependencies that we don't want bundled/are currently depending on the spark installation for. I did it this way because the shadowJar `exclude` filter does not let you exclude transitive dependencies, and I just wanted to exclude the entire spark/scala dependency tree.
- there was a problem where trying to run the tests kept giving me the "Could not find or load main class org.testng.TestNG" error, despite the class clearly being findable from the classpath I was providing. I added some excludes per this:
https://stackoverflow.com/questions/51455197/gradle-fatjar-could-not-find-or-load-main-class
(although I believe this is no longer strictly necessary after excluding all the transitive spark dependencies)

### CODE REVIEW

**Comment by cseed** [comment]
This is great, not having to enumerate the dependencies.

Hmm, this is potentially making the build 2x slower.

Your branch:

 > 17 | build_hail | Complete | Success 🎉 (0) | 8 minutes | log

A master deploy a few moments ago:

 > 16 | build_hail | Complete | 0 | 4 minutes | log

The cluster might have been under load when your tests run.  Can you do a bit of benchmarking to see if this is real?

**Comment by catoverdrive** [comment]
Hmm... shadowJar is building at about 2 minutes on my computer, 1 min 20s of which is compileScala. Master is building at about 1 min 40s (not sure how much scala compile is taking, forgot to check), and I managed to get it down to like 1 min 45s by not bundling some of the dependencies that we weren't bundling before.

**Comment by catoverdrive** [comment]
Should be back to 4 minutes now.

**Comment by cseed** [comment]
Indeed:

 > 17 | build_hail | Complete | Success 🎉 (0) | 4 minutes | log

Nice work!   I'll read over in more detail soon.

File: hail/generate-build-info.sh [hunk 1]
```diff
@@ -4,7 +4,7 @@ set -e
 
 REVISION=$(git rev-parse HEAD)
 DATE=$(date -u +%Y-%m-%dT%H:%M:%SZ)
-DATE=$(git rev-parse --abbrev-ref HEAD)
+BRANCH=$(git rev-parse --abbrev-ref HEAD)
```
Comment (cseed) [inline hail/generate-build-info.sh]: Good catch.

**Pull request merged**